### PR TITLE
feat(tasks): allow skipping tasks with the  flag

### LINKED ISSUE #1415: Ability to define safe/unsafe modes and force to ignore _tasks
### Actual Situation

If we define `_tasks`, the end user will need to use `--UNSAFE/--trust` to generate the project from the template.
In some situation, the `_tasks` commands are just convenience.

More precisely, in my case:
```yaml
_tasks:
    - "git init -b main && git add -A && git commit -m 'Initial commit'"
    - "pre-commit install"
```

### Desired Situation

I would like the end-user to be able to use the template without running any unsafe feature.

### Proposed solution

The end-user could be able to use a flag (e.g. `--no-script`) to not run any unsafe feature but still create the project.

From the template's creator side, an additional feature would be to mark the `_tasks` as `required/optional` to control if the end-user is allowed to copy the project without the tasks.

### PR DESCRIPTION
**Pull request by noirbizarre**
This PR add a `-T/--skip-tasks` flag allowing to skip template tasks execution.

Some known use cases I have:
- while testing my template, I want to be able to perform multiple rendering with the extensions actives but without running tasks (which may be slow or depending on connection...)
- when applying a template with some bootstrap tasks on an existing project, I want to be able to skip the bootstrap tasks because I know they will fail
- when applying a third party template that I trust, I want to only skip tasks for some reason (known to break, not compatible with the environment...)

Should fix #1415

### CODE REVIEW

File: copier/main.py
Comment (pawamoy) [inline copier/main.py:162]: Does not seem to belong to this PR?

**Review by pawamoy — COMMENTED** [review]
Looks solid :slightly_smiling_face: 

Do I understand correctly that passing `--skip-tasks` means I don't have to also pass `--trust`?

**Comment by noirbizarre** [comment]
No, you still have to pass `--trust` (if you want to have jinja extensions loaded), but if you have `--trust --skip-tasks` only extension will be loaded while `--trust` will load extension and execute tasks.
In fact it doesn't touch the `--trust` behavior, just skipping tasks, whether the template is trusted or not (but in the second case, tasks are disabled by security, not from user intent)

File: copier/main.py
Comment (pawamoy) [reply copier/main.py:162]: Ah, right :sweat_smile: Yes, please, another PR would be great (and immediately merged).

**Comment by pawamoy** [comment]
OK yes that makes sense.

File: copier/main.py
Comment (noirbizarre) [reply copier/main.py:162]: Done there: https://github.com/copier-org/copier/pull/1562 👍🏼 

I'll rebase this one onto it after merge

File: copier/cli.py
Comment (pawamoy) [inline copier/cli.py:161]: I still think `-T` would be better, or maybe `-K` (uppercase because that usually means "negative" option). @copier-org/maintainers any opinion on this?

Comment (sisp) [reply copier/cli.py:161]: `-T` seems best IMO, uppercase for negation and `T` for **t**asks (also in case we might have other skip flags in the future).

Comment (noirbizarre) [reply copier/cli.py]: PR updated 👍🏼

File: tests/test_migrations.py
Comment (pawamoy) [inline tests/test_migrations.py:58]: It's usually more readable to start with non-negated conditions:

```suggestion
    if skip_tasks:
        assert not (dst / "created-with-tasks.txt").exists()
        assert (dst / "delete-in-tasks.txt").exists()
    else:
        assert (dst / "created-with-tasks.txt").read_text() == "task 1\ntask 2\n"
        assert not (dst / "delete-in-tasks.txt").exists()

```

File: tests/test_migrations.py
Comment (pawamoy) [inline tests/test_migrations.py:83]: Same suggestion here:

```suggestion
    if skip_tasks:
        assert not (dst / "created-with-tasks.txt").exists()
        assert (dst / "delete-in-tasks.txt").exists()
    else:
        assert (dst / "created-with-tasks.txt").read_text() == "task 1\ntask 2\n" * 2
        assert not (dst / "delete-in-tasks.txt").exists()
```

**Review by pawamoy — COMMENTED** [review]
Two suggestions regarding readability and then I'll have nothing left to say I think :smile:

File: tests/test_migrations.py
Comment (noirbizarre) [reply tests/test_migrations.py:58]: I hesitated because depending on projects I contribute to, some prefer
- if non-negated else
- if nominal case else
I chose the 2nd because it was the existing case but I prefer the first one. Submitting the change ASAP

**Review by pawamoy — APPROVED** [review]
LGTM!

File: tests/test_tasks.py
Comment (sisp) [inline tests/test_tasks.py:63]: This test should also pass with

```suggestion
        unsafe=False,
```

right? I think we should test it that way to cover that `skip_tasks=True` makes `unsafe=True` unnecessary (when no other unsafe features are used – like here).

**Review by sisp — CHANGES_REQUESTED** [review]

File: tests/test_tasks.py
Comment (noirbizarre) [reply tests/test_tasks.py:63]: Yes it should but I don't see the point given that we already test that `unsafe=False` disable extensions, tasks and migrations.
That's exactly the point of this test: check that only tasks are skipped when the template is trusted and `--skip-tasks` is enabled. 
This change would just test that `--skip-tasks` don't do anything when the template is not trusted (eg. when tasks are already skipped).
The need to skip tasks only exists for trusted templates (eg. with extensions, tasks and migrations enabled), more precisely the need is to be able to run a template with extensions active but without the running tasks or to update with the extensions and migrations active but without the tasks. 
(Note that I took into account that it is planned that trusts will be persistent at some point, so this test should still pass when it is the case)

Comment (noirbizarre) [reply tests/test_tasks.py:63]: Edit: I just tried to parametrize `unsafe`, but it doesn't run for `unsafe=False`: trying to run a template with `tasks` and without `unsafe` raise `UnsafeTemplateError` (as expected by https://github.com/copier-org/copier/blob/master/tests/test_unsafe.py in fact)

Comment (sisp) [reply tests/test_tasks.py:63]: Ah, I missed the rationale you explained in https://github.com/copier-org/copier/pull/1561#issuecomment-2015137696. With that in mind, the behavior you observe is expected.

I wonder whether `--skip-tasks` should actually affect the `--trust` behavior. When you pass `--skip-tasks`, you avoid the security risk of tasks. So why should `copier copy --skip-tasks ...` fail because of unsafe features when the only unsafe feature of that template is "tasks"? I'm not sure that's intuitive. WDYT, @copier-org/maintainers?

**Review by sisp — CHANGES_REQUESTED** [review]
We also need to [document this setting here](https://github.com/copier-org/copier/blob/5fe373c83e7cc2810f23117c25b3d62f99396fc7/docs/configuring.md?plain=1#L1305).

File: tests/test_tasks.py
Comment (pawamoy) [reply tests/test_tasks.py:63]: @noirbizarre recalled previously that not only tasks are unsafe, but Jinja extensions too :slightly_smiling_face: 
Therefore `--skip-tasks` **should not** silently toggle `--trust` on.

Comment (sisp) [reply tests/test_tasks.py:63]: Yes, I didn't mean to toggle `--trust` but only ignore tasks when determining the unsafe features of a template when `--skip-trust` is provided.

I think you've convinced me to leave behavior as it is now. Also, we won't have `--skip-{jinja-extensions,migrations}` because skipping those would render the template useless. So considering `--skip-tasks` as a partial `--trust` seems conceptually wrong – adding to your reasoning towards the same conclusion.

**Comment by noirbizarre** [comment]
PR updated with the setting documentation 👍🏼

File: docs/configuring.md
Comment (pawamoy) [inline docs/configuring.md:1316]:
```suggestion
    It only skips tasks, not migration tasks.
```

File: docs/configuring.md
Comment (sisp) [inline docs/configuring.md:1318]: Sorry for being pedantic 🫣:

```suggestion
!!! question "Does it imply `--trust`?"
```

File: docs/configuring.md
Comment (sisp) [inline docs/configuring.md:1312]: I think UX would be a bit better with some links:

```suggestion
Skip template [tasks][tasks] execution, if set to `True`.
```

File: docs/configuring.md
Comment (sisp) [inline docs/configuring.md:1316]: I think UX would be a bit better with some links:

```suggestion
    It only skips [tasks][tasks], not [migration tasks][migrations].
```

File: docs/configuring.md
Comment (sisp) [inline docs/configuring.md:1320]: And I think UX would be a bit better with some links:

```suggestion
    This flag does not imply [`--trust`][unsafe], and will do nothing if not used with.
```

**Review by sisp — CHANGES_REQUESTED** [review]
Just some final :crossed_fingers: docs suggestions, @noirbizarre. I'll be happy to merge afterwards.

**Review by sisp — APPROVED** [review]
Fantastic, thanks @noirbizarre! :pray:

**Comment by yajo** [comment]
Sorry for being late to the party.

I think that, if you pass this flag, it should be taken into account around here to avoid marking the template as unsafe if it's only because of tasks but you're skipping them: https://github.com/copier-org/copier/blob/70aa69f7df430a611a20f940b47174da96b88dfa/copier/main.py#L233-L234

**Comment by sisp** [comment]
@yajo I think that's the same [thought that I had](https://github.com/copier-org/copier/pull/1561#discussion_r1537995998), but we decided against it in that discussion. WDYT?

**Comment by yajo** [comment]
Consider migrations. Even if the template uses migrations, we only mark it as unsafe if the current update will trigger them.

In the same fashion, if it has tasks but you're ignoring them, there's no unsafety produced by tasks. Other unsafety criteria should still be honored, but not that one.

**Pull request merged**
